### DISCUSSION TITLE
Tooltip overload

### DISCUSSION THREAD

**Discussion by Ekedev** [Live-Charts/LiveCharts2#354] [General]
For some reason when I enable tooltips it displays every value in the series and not the value I am hovering over.   I have a few thousand points in my series, but even when I zoom down to a level where the data should be granular I still see all the values in my series in the tooltip.

this is my LineSeries

chan0Series = new LineSeries<DateTimePoint>
            {
                Name = "Channel 0",
                Values = this._observableValuesChan0,
                Fill = null,
                GeometryFill = null,
                GeometryStroke = null,
                Stroke = new SolidColorPaint(chan0Color) { StrokeThickness = 2 },
                ScalesYAt = 0,
                LineSmoothness = 0,
                TooltipLabelFormatter = (chartPoint) => $"{chartPoint.PrimaryValue}"
            };

and the xaml

<lvc:CartesianChart x:Name="lvcHistChart" ZoomMode="X" ZoomingSpeed=".5" 
                            Series="{Binding Series}" XAxes="{Binding XAxes}" YAxes="{Binding YAxes}" TooltipPosition="Top" 
                                            LegendPosition="Right">                            
                        </lvc:CartesianChart>

**Comment by alexachso** [comment]
Hello, 
Did you find a solution for this? I am experiencing the same issue.
Thank you!

**Comment by jradio610** [comment]
Same issue here as well.

**Comment by langejay** [comment]
What is the value range in your data (x axis)?

**Comment by alexachso** [comment]
Hi,

Here is an actual example :) 
This is the chart I have.
![no tooltip](https://user-images.githubusercontent.com/12586000/159047367-73cc0063-4808-442a-b9e9-e7e88bcede9e.PNG)

![Oveload 2](https://user-images.githubusercontent.com/12586000/159047498-e56bbdf1-a724-424a-a879-6b95ad98dd52.png)

![Oveload 1](https://user-images.githubusercontent.com/12586000/159047526-f738b4cd-c524-4e5b-a401-f9b518c7bde9.png)

As you can see the tooltip shows me lots of points. I wouldn't mind that much if it was showing me all the Y points for the same X. (because in my case I have multiple Y points for the same X point).

And btw, when I zoom, the tooltip is all over the place... It even sometimes show it on my second screen.

**Reply by langejay** [reply]
you need to try to change the Unit Width for the chart.  As expected your values are <1 unit apart.  Try changing the unit width to something like 0.1 to 0.08 for your case. 

will be somthing like

Chart.XAxis[0].UnitWidth = 0.09;

Hope that helps!

__________________________________________________________________________________________________________________________

I am not sure if this is technically correct but I think this is used as a kind of 'detection band' to determine the 'breadth' of values to show in the tooltip.

I have had to develop an algorithm that TRIES to determine the UnitWidth based on '8 pixels'.  To do this however I have had to make some assumptions as the pixel width of the 'data area' is not provided.  Just the Width of the entire Cartesian plot.

It is something like:

    public override void UpdateUnitWidth()
    {
      double? min = XAxis.MinLimit;
      if (min == null)
      {
        List<double> mins = VM.GetXMin();
        if (mins.Count > 0) min = mins.Min();
      }

      double? max = XAxis.MaxLimit;
      if (max == null)
      {
        List<double> maxs = VM.GetXMax();
        if (maxs.Count > 0) max = maxs.Max();
      }

      if (min != null && max != null && min != max) {

        var pixelWidth = this.ActualWidth;
        if (pixelWidth > 0)
        {
          double dataWidth = (double)max - (double)min;
          double approxUnitsPerPixel = dataWidth / (pixelWidth - 70); // 70 approximate account for padding, axisy etc
          if (approxUnitsPerPixel < 1) XAxis.UnitWidth = 1;
          else XAxis.UnitWidth = approxUnitsPerPixel * 8;

        }
      }
      
    }
`

Note that the VM.Get... functions run a method that collects the min and max values for each series displayed in my View Model.  It is something you will need to alter/add for your own case.

Its crude, but it seems to work ok for me.  It is actioned when there is a zoom or the data is first drawn.

It would be good if the next release provided

 - the min and max of displayed series data without having to generate your own algorithm (when min/max limit is set to null).
 - The pixel width of the plot area (similar to ActualWidth but PlotAreaWidth for exapmple) so I don't have to guestimate.

**Reply by alexachso** [reply]
Thank you so much! Yes, it works!

I never thought about the unit width. My reason behind this was that if I configure the min and max to NULL (so it can autodetect the min and max) I assumed I don't need to worry about the unitwidth because I thought that the software can compute the unitwidth of the axis by applying the next formula by itself:

unitwidth = (max - min) / numberOfPointsInOneSeries; (The series that has the most number of points)

(I know it is not accurate... because you could have 2 points on the same X for example, but I guess it is better than assuming unitwidth always to 1). Your override function is even better than the simple formula I described above. I really like your approach on guesstimating the unitwidth. 

I am not complaining! I really like this library! If we can have a healthy discussion about on how to improve it a little bit why not do so? :)

Thank you again!

PS: I know it's not the thread for, but I observed that when I zoom it seems that the limiter factor on stop zooming is to have at least 6 drawn points of one series. Can I configure this limit of 6 drawn points to say 3 points or even 2 points instead so I can zoom a little bit further?
